src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java

```diff
@@ -855,6 +856,22 @@ private void handleOne(final Callback[] callbacks, final int idx) throws IOExcep
                     if (authenticationID != null) {
                         // always re-set the authentication name to ensure it hasn't changed.
                         setAuthenticationName(authenticationID);
+                    } else {
+                        // This is a special case to support scenarios where the identity was already established by some
+                        // external method (e.g.: EXTERNAL SASL and TLS) where only authorization is necessary. We delay authentication
+                        // until we receive an authorization request.
+                        // In the future, we may want to support external methods other than TLS peer authentication
+                        State state = stateRef.get();
+                        if (!(state instanceof NameAssignedState || state instanceof AuthorizedState)) {
```

Comment:
Yeah that's a good idea.  In this case a more general one might be needed though which is true if there is an assigned name _or_ there is an enclosing identity.